File: elixir-phoenix/lib/rocket/engine/writer.ex
defmodule Rocket.Engine.Writer do
  @moduledoc "SQL builders for INSERT/UPDATE and field validation."

  alias Rocket.Metadata.{Entity, Relation, Registry}

  defmodule RelationWrite do
    defstruct [:relation, :write_mode, data: []]
  end

  @doc "Build parameterized INSERT SQL. Returns {sql, params}."
  def build_insert_sql(entity, fields) do
    {cols, vals, params, _n} =
      entity.fields
      |> Enum.reduce({[], [], [], 0}, fn f, {cols, vals, params, n} ->
        cond do
          f.name == entity.primary_key.field && entity.primary_key.generated ->
            {cols, vals, params, n}

          f.auto in ["create", "update"] ->
            {cols, vals, params, n}

          f.name == "deleted_at" ->
            {cols, vals, params, n}

          true ->
            val = Map.get(fields, f.name)

            cond do
              val != nil ->
                n = n + 1
                {cols ++ [f.name], vals ++ ["$#{n}"], params ++ [val], n}

              f.default != nil ->
                n = n + 1
                {cols ++ [f.name], vals ++ ["$#{n}"], params ++ [f.default], n}

              true ->
                {cols, vals, params, n}
            end
        end
      end)

    # Add auto-timestamp fields
    {cols, vals} =
      Enum.reduce(entity.fields, {cols, vals}, fn f, {cols, vals} ->
        if f.auto in ["create", "update"] do
          {cols ++ [f.name], vals ++ ["NOW()"]}
        else
          {cols, vals}
        end
      end)

    sql =
      "INSERT INTO #{entity.table} (#{Enum.join(cols, ", ")}) VALUES (#{Enum.join(vals, ", ")}) RETURNING #{entity.primary_key.field}"

    {sql, params}
  end

  @doc "Build parameterized UPDATE SQL. Returns {sql, params} or {\"\", []} if nothing to update."
  def build_update_sql(entity, id, fields) do
    updatable = Entity.updatable_fields(entity)

    {sets, params, n} =
      Enum.reduce(updatable, {[], [], 0}, fn f, {sets, params, n} ->
        case Map.fetch(fields, f.name) do
          {:ok, val} ->
            n = n + 1
            {sets ++ ["#{f.name} = $#{n}"], params ++ [val], n}

          :error ->
            {sets, params, n}
        end
      end)

    # Auto-update timestamp
    sets =
      Enum.reduce(entity.fields, sets, fn f, sets ->
        if f.auto == "update" do
          sets ++ ["#{f.name} = NOW()"]
        else
          sets
        end
      end)

    if sets == [] do
      {"", []}
    else
      n = n + 1
      params = params ++ [id]
      where = "#{entity.primary_key.field} = $#{n}"

      where =
        if entity.soft_delete do
          where <> " AND deleted_at IS NULL"
        else
          where
        end

      sql = "UPDATE #{entity.table} SET #{Enum.join(sets, ", ")} WHERE #{where}"
      {sql, params}
    end
  end

  @doc "Validate incoming fields against entity metadata."
  def validate_fields(entity, fields, is_create) do
    errs = []

    # Check for unknown fields
    errs =
      Enum.reduce(Map.keys(fields), errs, fn key, errs ->
        if Entity.has_field?(entity, key) do
          errs
        else
          errs ++ [%{field: key, rule: "unknown", message: "Unknown field: #{key}"}]
        end
      end)

    # Check required fields on create
    errs =
      if is_create do
        Enum.reduce(Entity.writable_fields(entity), errs, fn f, errs ->
          if f.required && !f.nullable do
            val = Map.get(fields, f.name)

            if val == nil || val == "" do
              errs ++ [%{field: f.name, rule: "required", message: "#{f.name} is required"}]
            else
              errs
            end
          else
            errs
          end
        end)
      else
        errs
      end

    # Check enum constraints
    errs =
      Enum.reduce(entity.fields, errs, fn f, errs ->
        if f.enum != nil && f.enum != [] do
          val = Map.get(fields, f.name)

          if val != nil do
            str_val = to_string(val)

            if str_val in f.enum do
              errs
            else
              errs ++
                [
                  %{
                    field: f.name,
                    rule: "enum",
                    message: "#{f.name} must be one of: #{Enum.join(f.enum, ", ")}"
                  }
                ]
            end
          else
            errs
          end
        else
          errs
        end
      end)

    errs
  end

  @doc "Split request body into entity fields and relation writes."
  def separate_fields_and_relations(entity, registry, body) do
    Enum.reduce(body, {%{}, %{}, []}, fn {key, val}, {fields, rel_writes, unknown} ->
      cond do
        Entity.has_field?(entity, key) ->
          {Map.put(fields, key, val), rel_writes, unknown}

        true ->
          rel = Registry.find_relation_for_entity(registry, key, entity.name)

          if rel != nil && rel.source == entity.name do
            case parse_relation_write(rel, val) do
              {:ok, rw} ->
                {fields, Map.put(rel_writes, key, rw), unknown}

              :error ->
                {fields, rel_writes, unknown ++ [key]}
            end
          else
            {fields, rel_writes, unknown ++ [key]}
          end
      end
    end)
  end

  defp parse_relation_write(rel, val) when is_map(val) do
    write_mode = Map.get(val, "_write_mode", Relation.default_write_mode(rel))

    case Map.get(val, "data") do
      data when is_list(data) ->
        items = Enum.filter(data, &is_map/1)

        {:ok,
         %RelationWrite{
           relation: rel,
           write_mode: write_mode,
           data: items
         }}

      _ ->
        :error
    end
  end

  defp parse_relation_write(_, _), do: :error
end

File: expressjs/src/engine/writer.ts
import { randomUUID } from "node:crypto";
import type { Entity, Relation } from "../metadata/types.js";
import type { Registry } from "../metadata/registry.js";
import {
  hasField,
  writableFields,
  updatableFields,
  defaultWriteMode,
} from "../metadata/types.js";
import { ParamBuilder } from "./query.js";
import type { ErrorDetail } from "./errors.js";
import { getDialect } from "../store/postgres.js";

export interface RelationWrite {
  relation: Relation;
  writeMode: string;
  data: Record<string, any>[];
}

export function buildInsertSQL(
  entity: Entity,
  fields: Record<string, any>,
): [string, any[]] {
  const pb = new ParamBuilder();
  const cols: string[] = [];
  const vals: string[] = [];

  const dialect = getDialect();
  const needsAppUUID =
    entity.primary_key.generated &&
    entity.primary_key.type === "uuid" &&
    dialect.name() === "sqlite";

  for (const f of entity.fields) {
    if (f.name === entity.primary_key.field && entity.primary_key.generated) {
      if (needsAppUUID) {
        // SQLite has no gen_random_uuid() default — generate in app code
        cols.push(f.name);
        vals.push(pb.add(randomUUID()));
      }
      continue;
    }
    if (f.auto === "create" || f.auto === "update") {
      continue;
    }
    if (f.name === "deleted_at") {
      continue;
    }

    let val = fields[f.name];
    if (val === undefined) {
      if (f.default != null) {
        val = f.default;
      } else if (!f.required) {
        continue;
      } else {
        continue;
      }
    }
    cols.push(f.name);
    vals.push(pb.add(val));
  }

  // Add auto-timestamp fields
  for (const f of entity.fields) {
    if (f.auto === "create" || f.auto === "update") {
      cols.push(f.name);
      vals.push("NOW()");
    }
  }

  const sql = `INSERT INTO ${entity.table} (${cols.join(", ")}) VALUES (${vals.join(", ")}) RETURNING ${entity.primary_key.field}`;
  return [sql, pb.params];
}

export function buildUpdateSQL(
  entity: Entity,
  id: any,
  fields: Record<string, any>,
): [string, any[]] | [string, null] {
  const pb = new ParamBuilder();
  const sets: string[] = [];

  for (const f of updatableFields(entity)) {
    const val = fields[f.name];
    if (val === undefined) continue;
    sets.push(`${f.name} = ${pb.add(val)}`);
  }

  // Auto-update timestamp
  for (const f of entity.fields) {
    if (f.auto === "update") {
      sets.push(`${f.name} = NOW()`);
    }
  }

  if (sets.length === 0) {
    return ["", null];
  }

  let where = `${entity.primary_key.field} = ${pb.add(id)}`;
  if (entity.soft_delete) {
    where += " AND deleted_at IS NULL";
  }

  const sql = `UPDATE ${entity.table} SET ${sets.join(", ")} WHERE ${where}`;
  return [sql, pb.params];
}

export function buildSoftDeleteSQL(
  entity: Entity,
  id: any,
): [string, any[]] {
  const sql = `UPDATE ${entity.table} SET deleted_at = NOW() WHERE ${entity.primary_key.field} = $1 AND deleted_at IS NULL`;
  return [sql, [id]];
}

export function buildHardDeleteSQL(
  entity: Entity,
  id: any,
): [string, any[]] {
  const sql = `DELETE FROM ${entity.table} WHERE ${entity.primary_key.field} = $1`;
  return [sql, [id]];
}

export function validateFields(
  entity: Entity,
  fields: Record<string, any>,
  isCreate: boolean,
): ErrorDetail[] {
  const errs: ErrorDetail[] = [];

  // Check for unknown fields
  for (const key of Object.keys(fields)) {
    if (!hasField(entity, key)) {
      errs.push({
        field: key,
        rule: "unknown",
        message: `Unknown field: ${key}`,
      });
    }
  }

  if (isCreate) {
    for (const f of writableFields(entity)) {
      if (f.required && !f.nullable) {
        const val = fields[f.name];
        if (val === undefined || val === null || val === "") {
          errs.push({
            field: f.name,
            rule: "required",
            message: `${f.name} is required`,
          });
        }
      }
    }
  }

  // Check enum constraints
  for (const f of entity.fields) {
    if (!f.enum || f.enum.length === 0) continue;
    const val = fields[f.name];
    if (val === undefined || val === null) continue;
    const strVal = String(val);
    if (!f.enum.includes(strVal)) {
      errs.push({
        field: f.name,
        rule: "enum",
        message: `${f.name} must be one of: ${f.enum.join(", ")}`,
      });
    }
  }

  return errs;
}

export function separateFieldsAndRelations(
  entity: Entity,
  registry: Registry,
  body: Record<string, any>,
): [Record<string, any>, Map<string, RelationWrite>, string[]] {
  const fields: Record<string, any> = {};
  const relWrites = new Map<string, RelationWrite>();
  const unknownKeys: string[] = [];

  for (const [key, val] of Object.entries(body)) {
    if (hasField(entity, key)) {
      fields[key] = val;
      continue;
    }

    const rel = registry.findRelationForEntity(key, entity.name);
    if (rel && rel.source === entity.name) {
      const rw = parseRelationWrite(rel, val);
      if (rw) {
        relWrites.set(key, rw);
        continue;
      }
    }

    unknownKeys.push(key);
  }

  return [fields, relWrites, unknownKeys];
}

function parseRelationWrite(
  rel: Relation,
  val: any,
): RelationWrite | null {
  if (typeof val !== "object" || val === null || Array.isArray(val)) {
    return null;
  }

  const rw: RelationWrite = {
    relation: rel,
    writeMode: defaultWriteMode(rel),
    data: [],
  };

  if (typeof val._write_mode === "string") {
    rw.writeMode = val._write_mode;
  }

  const data = val.data;
  if (!Array.isArray(data)) {
    return null;
  }

  for (const item of data) {
    if (typeof item === "object" && item !== null && !Array.isArray(item)) {
      rw.data.push(item);
    }
  }

  return rw;
}
